x-pack/plugins/osquery/public/packs/queries/ecs_mapping_editor_field.tsx

```diff
@@ -721,217 +723,258 @@ interface OsqueryColumn {
   index: boolean;
 }
 
-export const ECSMappingEditorField = React.memo(
-  ({ euiFieldProps }: ECSMappingEditorFieldProps) => {
-    const { trigger } = useFormContext();
-    const { fields, append, remove } = useFieldArray<{ ecs_mapping: EcsMappingFormField[] }>({
-      name: 'ecs_mapping',
-    });
+export const ECSMappingEditorField = React.memo(({ euiFieldProps }: ECSMappingEditorFieldProps) => {
+  const {
+    setError,
+    clearErrors,
+    watch: watchRoot,
+    register: registerRoot,
+    setValue: setValueRoot,
+  } = useFormContext<{ query: string; ecs_mapping: ECSMapping }>();
+  const [query, ecsMapping] = watchRoot(['query', 'ecs_mapping'], { ecs_mapping: {} });
+
+  registerRoot('ecs_mapping');
+
+  const { control, trigger, watch, formState, resetField, getFieldState } = useForm<{
+    ecsMappingArray: ECSMappingArray;
+  }>({
+    mode: 'all',
+    shouldUnregister: true,
+    defaultValues: {
+      ecsMappingArray: !isEmpty(convertECSMappingToArray(ecsMapping))
+        ? [...convertECSMappingToArray(ecsMapping), defaultEcsFormData]
+        : [defaultEcsFormData],
+    },
+  });
+  const { fields, append, remove, replace } = useFieldArray({
+    control,
+    name: 'ecsMappingArray',
+  });
+
+  const formValue = watch();
+  const ecsMappingArrayState = getFieldState('ecsMappingArray', formState);
+  const [osquerySchemaOptions, setOsquerySchemaOptions] = useState<OsquerySchemaOption[]>([]);
 
-    const itemsList = useRef<Array<{ id: string }>>([]);
-    const [osquerySchemaOptions, setOsquerySchemaOptions] = useState<OsquerySchemaOption[]>([]);
-    const { query, ...formData } = useWatch() as unknown as {
-      query: string;
-      ecs_mapping: EcsMappingFormField[];
-    };
-
-    useEffect(() => {
-      // Additional 'suspended' validation of osquery ecs fields. fieldsToValidateOnChange doesn't work because it happens before the osquerySchema gets updated.
-      const fieldsToValidate = prepareEcsFieldsToValidate(fields);
-      // it is always at least 2 - empty fields
-      if (fieldsToValidate.length > 2) {
-        setTimeout(() => trigger('ecs_mapping'), 0);
+  useEffect(() => {
+    const subscription = watchRoot((data, payload) => {
+      if (payload.name === 'ecs_mapping') {
+        const parsedMapping = convertECSMappingToObject(formValue.ecsMappingArray);
+        if (!deepEqual(data.ecs_mapping, parsedMapping)) {
```

Comment:
this seem very complex, could you shed some light on why we need this please?

Reply:
here we are syncing internal `ecsMappingArray` state with the proper `ecs_mapping` field value, so that way we always store an object in `ecs_mapping` and we avoid converting it whole the time and in `ecsMappingArray` we store the "draft" state where we make changes to